File: src/main/java/thkoeln/dungeon/player/domain/Player.java
package thkoeln.dungeon.player.domain;


import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.beans.factory.annotation.Value;
import thkoeln.dungeon.game.domain.Game;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Entity
@Setter
@Getter
@NoArgsConstructor
public class Player {
    @Id
    private final UUID id = UUID.randomUUID();

    private String name;
    private String email;
    private UUID bearerToken;

    @OneToMany ( cascade = { CascadeType.MERGE, CascadeType.REMOVE }, fetch = FetchType.EAGER )
    private final List<GameParticipation> gameParticipations = new ArrayList<>();


    /**
     * Choose a random and unique name and email for the player
     */
    public void assignRandomName() {
        String randomNickname = NameGenerator.generateName();
        setName( randomNickname );
        setEmail( randomNickname + "@microservicedungeon.com" );
    }

    public boolean isReadyToPlay() {
        return ( bearerToken != null );
    }


    public void participateInGame( Game game ) {
        GameParticipation gameParticipation = new GameParticipation( game );
        gameParticipations.add( gameParticipation );
    }


    public boolean isParticipantInGame( Game game ) {
        return ( findParticipationFor( game ) != null );
    }


    private GameParticipation findParticipationFor( Game game ) {
        Optional<GameParticipation> found = gameParticipations.stream()
                        .filter( gp -> gp.getGame().equals( game ) ).findFirst();
        return found.isPresent() ? found.get() : null;
    }



    public void playRound() {
        // todo
    }

    @Override
    public String toString() {
        return "Player '" + name + "' (" + email + "), bearerToken: " + bearerToken;
    }
}

File: src/main/java/thkoeln/dungeon/player/domain/Player.kt
package thkoeln.dungeon.player.domain

import lombok.*
import thkoeln.dungeon.game.domain.Game
import java.util.*
import javax.persistence.*

@Entity
@Setter
@Getter
@NoArgsConstructor
class Player {
    @Id
    private val id = UUID.randomUUID()
    var name: String? = null
    var email: String? = null
    var bearerToken: UUID? = null

    @OneToMany(cascade = [CascadeType.MERGE, CascadeType.REMOVE], fetch = FetchType.EAGER)
    private val gameParticipations: MutableList<GameParticipation> = ArrayList()

    /**
     * Choose a random and unique name and email for the player
     */
    fun assignRandomName() {
        val randomNickname = NameGenerator.generateName()
        this.name = randomNickname
        this.email = "$randomNickname@microservicedungeon.com"
    }

    val isReadyToPlay: Boolean
        get() = bearerToken != null

    fun participateInGame(game: Game?) {
        val gameParticipation = GameParticipation(game)
        gameParticipations.add(gameParticipation)
    }

    fun isParticipantInGame(game: Game): Boolean {
        return findParticipationFor(game) != null
    }

    private fun findParticipationFor(game: Game): GameParticipation? {
        val found = gameParticipations.stream()
            .filter { gp: GameParticipation -> gp.game == game }.findFirst()
        return if (found.isPresent) found.get() else null
    }

    fun playRound() {
        // todo
    }

    override fun toString(): String {
        return "Player '$name' ($email), bearerToken: $bearerToken"
    }
}